internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,178 @@
+package addonsdk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	statusReporterSingleton *StatusReporter
+	initializeSingletonOnce sync.Once
+)
+
+// a nice heartbeat-reporter implemented by the MT-SRE which our tenants can use
+// if they don't like it, they can implement their own heartbeat reporter by creating a type which implements the `addonsdk.statusReporterClient` interface
+type StatusReporter struct {
+	// object provided by the client/tenants which implements the addonsdk.client interface
+	addonInstanceInteractor client
+	addonName               string
+	addonTargetNamespace    string
+
+	// the latest conditions which the heartbeat reporter would be reporting periodically
+	latestConditions []metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	interval time.Duration
+	ticker   *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	updateCh chan updateOptions
+
+	//for tracking if the heartbeat reporter is running or done running
+	doneCh chan bool
+
+	log logr.Logger
+}
+
+// SetupStatusReporter sets up a singleton of the type `statusReporter` (only if it doesn't exist yet) and returns it to the caller.
+func SetupStatusReporter(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) *StatusReporter {
+	initializeSingletonOnce.Do(func() {
+		statusReporterSingleton = &StatusReporter{
+			addonInstanceInteractor: addonInstanceInteractor,
+			addonName:               addonName,
+			addonTargetNamespace:    addonTargetNamespace,
+			latestConditions: []metav1.Condition{
+				{
+					Type:    AddonHealthyConditionType,
+					Status:  metav1.ConditionUnknown,
+					Reason:  "NoHealthReported",
+					Message: fmt.Sprintf("Addon %q hasn't reported health yet", addonName),
+				},
+			},
+			updateCh: make(chan updateOptions),
+			doneCh:   make(chan bool),
+			log:      logger,
+		}
+		// because the heartbeat reporter still hasn't been started
+		defer close(statusReporterSingleton.doneCh)
+	})
+	return statusReporterSingleton
+}
+
+func (sr *StatusReporter) Start(ctx context.Context) error {
+	// ensures to tie only one heartbeat-reporter loop at a time to a StatusReporter object
+	select {
+	case <-sr.doneCh:
+		sr.log.Info("StatusReporter already found to be running.")
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	sr.doneCh = make(chan bool)
```

Comment:
Yes you're right. Though now as per https://github.com/openshift/reference-addon/pull/9#discussion_r815863273 the `sr.doneCh = make(chan bool)` (and the subsequent lines) would come under `case <-sr.doneCh` thereby making that section concurrency safe.

Reply:
Converted sr.doneCh to a buffered channel for making concurrency-safe starts of the StatusReporter. Check this out - https://github.com/openshift/reference-addon/pull/9/commits/4ca04339f1dc32201927525016e76947e929234e

```go
	if len(sr.doneCh) == 0 {
		return fmt.Errorf("can't report AddonInstance spec change on a stopped StatusReporter")
	}
	select {
	case sr.updateCh <- updateOptions{addonInstance: newAddonInstance}:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
```

Although, technically as you can see, the `ReportAddonInstanceSpecChange`, `SetConditions` are still not concurrency-safe because the first `if` clause checking `len(doneCh) == 0` might get bypassed when instantly (and concurrently) the StatusReporter stops just after only the `if` clause executes. Meaning, in this situation, status reporter would be stopped, yet the `select` section would execute. But if that happens too, the select section will be blocked until either `ctx.Done()` happens or concurrently a new Status Reporter loop starts, which is going to be a  receiver for `sr.updateCh`, thereby triggering the `case sr.updateCh <- updateOptions{addonInstance: newAddonInstance}:`.

So should we be okay with that or add better/more synchronization to this to ensure that this doesn't happen ?

Reply:
PS: the strategy in my head for the latter requirement is to make use of RWMutex for efficiently locking the doneCh whenever it's accessed.

`ReferenceAddonInstanceSpecChange` and `SetConditions` would RLock it.
And `Start` would `Lock()` it and instantly `Unlock()` it as soon as `case sr.doneCh <- true` executes or `case ctx.Done()` executes.

Reply:
Let me make a quick commit to depict that more clearly :)

Reply:
There you go - https://github.com/openshift/reference-addon/pull/9/commits/8695d6d8ae0e4ef54f327f886425a6ae79ecded7 :)